src/network_helper_win.rs

```diff
@@ -0,0 +1,109 @@
+}
+
+impl Iterator for InterfaceAddressIterator {
+    type Item = (String, MacAddr);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.adapter.is_null() {
+            return None;
+        }
+        unsafe {
+            let adapter = self.adapter;
+            // Move to the next adapter
+            self.adapter = (*adapter).Next;
+            if let Ok(interface_name) = u16_ptr_to_string((*adapter).FriendlyName).into_string() {
+                // take the first 6 bytes and return the MAC address instead
+                let [mac @ .., _, _] = (*adapter).PhysicalAddress;
+                Some((interface_name, MacAddr::from(mac)))
+            } else {
+                // Not sure whether error can occur when parsing adapter name.
+                self.next()
+            }
+        }
+    }
+}
+
+impl Drop for InterfaceAddressIterator {
+    fn drop(&mut self) {
+        unsafe {
+            libc::free(self.buf as _);
+        }
+    }
+}
+
+pub(crate) fn get_interface_address() -> Result<InterfaceAddressIterator, String> {
+    unsafe {
+        // https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses#remarks
+        // A 15k buffer is recommended
+        let mut size: u32 = 15 * 1024;
+        let mut buf = null_mut();
+        let mut ret = ERROR_SUCCESS;
+
+        // https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses#examples
+        // Try to retrieve adapter information up to 3 times
+        for _ in 0..3 {
+            buf = libc::malloc(size as usize) as PIP_ADAPTER_ADDRESSES;
+            if buf.is_null() {
+                // insufficient memory available
+                // https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/malloc?view=msvc-170#return-value
+                // malloc is not documented to set the last-error code
+                return Err("failed to allocate memory for IP_ADAPTER_ADDRESSES".to_string());
+            }
+
+            ret = GetAdaptersAddresses(
+                AF_UNSPEC as u32,
+                GAA_FLAG_SKIP_MULTICAST | GAA_FLAG_SKIP_ANYCAST | GAA_FLAG_SKIP_DNS_SERVER,
+                null_mut(),
+                buf,
+                &mut size,
+            );
+
+            if ret == ERROR_BUFFER_OVERFLOW {
+                // if the given memory size is too small to hold the adapter information,
+                // the SizePointer returned will point to the required size of the buffer,
+                // and we should continue.
+                libc::free(buf as _);
+                buf = null_mut();
+            } else {
+                // Otherwise, break the loop and check the return code again
+                break;
+            }
+        }
+
+        if ret == ERROR_SUCCESS && !buf.is_null() {
+            Ok(InterfaceAddressIterator { buf, adapter: buf })
+        } else {
+            Err(format!("GetAdaptersAddresses() failed with code {}", ret))
```

Comment:
You need to free `buf` if it's not NULL in here. I think the best solution would instead to wrap it into a type which implements `Drop`.